File: src/februaryreset/graph/BFS.java
package februaryreset.graph;

import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;
import java.util.Stack;

public class BFS {

    private boolean[] visited;
    private Integer[] edgeTo;

    private Integer source;

    private Graph graph;

    public BFS(Graph graph, Integer source) {
        this.graph = graph;
        this.source = source;
        this.visited = new boolean[graph.V()];
        this.edgeTo = new Integer[graph.E()];
    }

    public void search() {
        Queue<Integer> todo = new LinkedList<>();
        todo.add(this.source);
        this.visited[this.source] = true;

        while (!todo.isEmpty()) {
            Integer polled = todo.poll();
            Set<Integer> neighbors = this.graph.adjList.get(polled);
            for (Integer n : neighbors) {
                if (!this.visited[n]) {
                    this.visited[n] = true;
                    edgeTo[n] = polled;
                    todo.add(n);
                }
            }
        }
    }

    public boolean hasPath(Integer t) {
        return visited[t];
    }

    public Stack<Integer> pathTo(int t) {
        Stack<Integer> path = new Stack<>();
        for (int x = this.source; x != t; x = edgeTo[t])
            path.push(x);

        return path;
    }
}

File: python-code/graphs/bfs.py
from typing import *
from base_graph import BaseGraph


class BFS(BaseGraph):
    def __init__(self,
                 graph: Dict[int, Set[int]],
                 origin: int):
        super().__init__(graph, origin)
        self._bfs(self.origin)

    def _bfs(self, node: int):
        queue: List[int] = [node]

        while queue:
            level_size: int = len(queue)
            for _ in range(level_size):
                polled: int = queue.pop()
                self.visited.add(polled)

                neighbors: Set[int] = self.graph[polled]
                for n in neighbors:
                    if n not in self.visited:
                        self.edge_to[n] = polled
                        queue.append(n)


if __name__ == '__main__':
    graph: Dict[int, Set[int]] = {
        1: {2, 4},
        2: {1, 3},
        3: {2},
        4: {1, 5},
        5: {4}
    }

    graph_bfs: BFS = BFS(graph, 1)
    print(graph_bfs.path_to(5))
